Document transaction fees (#288)
substrate-developer-hub/substrate-developer-hub.github.io: docs/development/module/fees.md

---
title: Transaction Fees
---

When transactions are submitted to a blockchain, they are executed by the nodes
in the network. To be economically sustainable, nodes charge a fee to execute a
transaction that must be covered by the submitter of the transaction. The cost
to execute transactions could vary over orders of magnitude, and thus Substrate
provides a flexible mechanism for characterizing the total, presumably minimum,
cost of a transaction in order to be included into a block.

> Make sure to read [Weights] before implementing fees.

Author: Incomplete link

  [`TransactionBaseFee`](https://crates.parity.io/srml_transaction_payment/trait.Trait.html#associatedtype.TransactionBaseFee)
* `length_fee` a per-byte fee that is multiplied by the length, in bytes, of the
  encoded transaction. See
  [`TransactionByteFee`](https://crates.parity.io/srml_transaction_payment/trait.Trait.html#associatedtype.TransactionByteFee)
* `weight_fee` a per-weight-unit fee that is multiplied by the weight of the
  transaction. As mentioned, weight of each dispatch is denoted via the flexible
  `#[weight]` annotation. Knowing the weight, it must be converted to a
  deductible `balance` type (typically denoted by a module that implements
  `Currency`, `srml-balances` in substrate node). For this, each runtime must
  define a
  [`WeightToFee`](https://crates.parity.io/srml_transaction_payment/trait.Trait.html#associatedtype.WeightToFee)
  type that makes the conversion.

based on the above, the final fee of a dispatchable is:

```
fee =
  base_fee +
  len(tx) * length_fee +
  WeightToFee(weight)
```

TODO: Examples

## Substrate Default Fee System

How to use the base fee system

### Declaring a Trasnaction Weight

### Changing the Default Weight

> 

Author: Should this be covered in the weight article?

### Changing the fee calculation

### Using Operational Transactions

### Adjusting Multiplier
The above formula gives a fee which is __logically constant through time__. Of
course, the weight can, in practice, be dynamic and based on what `WeightToFee`
is defined to be, the final fee can have some degree of variability. As for the
length fee, the inputs of the transaction could change the length and hence
affecting the length fee. Nonetheless, these changes are independent and a
_general update logic to the entire fee cannot be composed out of them
trivially_. In other words, for any dispatchable, given the same inputs, _it
will always incur the same cost_. This might not always be desirable. Chains

> might need to increase or decrease fees based on sime condition.  To fulfill

Author: typo sime -> some

this requirement, Substrate provides:
  - a multiplier stored in the transaction-payment module that is applied to the
    outcome of the above formula by default (needless to say, the default value
    of which is _multiplication identity_, meaning that it has no effect). This
    is stored in
    [`NextFeeMultiplier`](https://crates.parity.io/srml_transaction_payment/struct.Module.html#method.next_fee_multiplier)
    storage.
  - a configurable parameter for a runtime to describe how this multiplier can
    change. This is expressed via
    [`FeeMultiplierUpdate`](https://crates.parity.io/srml_transaction_payment/trait.Trait.html#associatedtype.FeeMultiplierUpdate)

## Building a Custom Fee System

> ## Gluing it Together: How Things Work in Substrate

Author: This title is uses "Substrate" to refer to the concrete node that lives in paritytech/substrate/node. We often use "Substrate" to refer to the framework. Should we clarify what we mean here? I often speak of "the default Substrate node".

Reviewer: I actually removed this and just briefly hint about the default in the text.

[weight](https://crates.parity.io/sr_primitives/weights/index.html) system.
Hence, we will first briefly describe how weights work.

As mentioned, weights can technically refer to, or represent, numerous _limited_
resources. A custom implementation may use complex structures to demonstrate
this. At the time of this writing, substrate weights are simply a [numeric
value](https://crates.parity.io/sr_primitives/weights/type.Weight.html). Each
dispatchable function is given a weight using the `#[weight = $x]` annotation,
where `$x` is some function capable of determining the weight of the dispatch.
`$x` can access and examine the arguments of the dispatch. Nonetheless, a weight
calculation should always:
- Be computable __ahead of dispatch__. A block producer, or a validator, should
  be able to examine the weight of a dispatchable before actually deciding to
  accept it or not.
- While not enforced, calculating the weight should not consume much resource
  itself. It does not make much sense to spend almost the same amount of time as
  dispatching a transaction to estimate how much weight it will consume. Thus,
  weight computation should typically be very fast, much faster than

>   dispatching. This is analogous to a `poll()` function of the `Future` trait.

Reviewer 2: I am not really a fan of bringing up `poll` and future here. I think your statement is clear without this example.

respective module.

```rust
use sr_primitives::{traits::Convert, weights::Weight}
// Assume this is the balance type
type Balance = u64;

// Assume we want all the weights to have a `100 + 2 * w` conversion to fees
struct CustomWeightToFee;
impl Convert<Weight, Balance> for CustomWeightToFee {
    fn convert(w: Weight) -> Balance {
        let a = Balance::from(100);
        let b = Balance::from(2);
        let w = Balance::from(w);
        a + b * w
    }
}

parameter_types! {
    TransactionBaseFee: Balance = 10;
    TransactionByteFee: Balance = 10;
}

impl transaction_payment::Trait {
    TransactionBaseFee = TransactionBaseFee;
    TransactionByteFee = TransactionByteFee;
    WeightToFee = CustomWeightToFee;
    // we will get to this one soon enough
    FeeMultiplierUpdate = ();
}

```

Two further questions need to be answered, given the above snippet.

1. What is the purpose of `FeeMultiplierUpdate`?
2. Knowing how we are capable of defining the conversion of weights to fees (via `WeightToFee`), the
   question remains, how can we define and customize the weights in the first place?

> We will answer each question in the following sections, respectively.

Reviewer 2: I really don't like this style of writing, sorry. You should get a second opinion though.

Reviewer: Appreciated. I read it again and I liked it tbh. Don't know what to do but open to edit.

## Adjusting Multiplier
The above formula gives a fee which is __logically constant through time__. Of course, the weight
can be dynamic and based on what `WeightToFee` is defined to be, the final fee can have some degree
of variability. As for the length fee, the inputs of the transaction could change the length and
hence affecting the length fee. Nonetheless, these changes are independent and a _general update
logic to the **entire fee** cannot be composed out of them trivially_. In other words, for any
dispatchable, given the same inputs, _it will always incur the same cost_. This might not always be
desirable. Chains might need to increase or decrease fees based on some condition.  To fulfill this
requirement, Substrate provides:
  - a multiplier stored in the transaction-payment module that is applied to the outcome of the

>     above formula by default (needless to say, the default value of which is _multiplication

Reviewer 2: ```suggestion
    above formula by default (needless to say, the default value of which is _multiplicative
```

Is this comment necessary?

Reviewer: I assume yes: to emphasise that the whole thing has a switch btn.

    identity_, meaning that it has no effect). This is stored in
    [`NextFeeMultiplier`](/rustdocs/master/srml_transaction_payment/struct.Module.html#method.next_fee_multiplier)
    storage and can be configured through the genesis spec of the module.
  - a configurable parameter for a runtime to describe how this multiplier can change. This is
    expressed via
    [`FeeMultiplierUpdate`](/rustdocs/master/srml_transaction_payment/trait.Trait.html#associatedtype.FeeMultiplierUpdate).

> `NextFeeMultiplier` has the type `Fixed64`, which can represent a fixed point number with a billion

Reviewer 2: ```suggestion
`NextFeeMultiplier` has the type `Fixed64`, which can represent a fixed point number with accuracy to a billionth
```

Is the billionth comment necessary?

Reviewer: yeah not important here.

points accuracy. So, given the final fee formula above, the final version would be:

```
fee =
  base_fee +
  len(tx) * length_fee +
  WeightToFee(weight)

final_fee = fee * NextFeeMultiplier
```

Updating the `NextFeeMultiplier` has a similar manner as `WeightToFee`. The
[`FeeMultiplierUpdate`](/rustdocs/master/srml_transaction_payment/trait.Trait.html#associatedtype.FeeMultiplierUpdate)
associated type in `transaction-payment` is defined as a `Convert<Fixed64, Fixed64>`, which should

> be read as: it receives the previous multiplier and spits out the next one.

Reviewer 2: ```suggestion
be read: "it receives the previous multiplier and spits out the next one".
```

The default update function is inspired by the Polkadot network and implements a targeted adjustment
in which a target saturation level of block weight is defined. If the previous block is more

> saturated, then the fees are slightly increases. Similarly, if less transaction than the target are

Reviewer 2: ```suggestion
saturated, then the fees are slightly increased. Similarly, if less transactions than the target are
```

in the previous block, fees are decreased by a small amount. More information about this can be

> found in the [web3 research

Reviewer 2: ```suggestion
found in the [Web3 research
```

page](https://research.web3.foundation/en/latest/polkadot/Token%20Economics/#relay-chain-transaction-fees).

## Substrate Weight System

> > This section assumes that you have already read the `Weight` section

Reviewer 2: ```suggestion
> This section assumes that you have already read the [conceptual weight document](conceptual/runtime/weight.md).
```

> The entire substrate runtime module library is already annotated with a simple and fixed weight

Reviewer 2: ```suggestion
The entire Substrate Runtime Module Library is already annotated with a simple and fixed weight
```

system. A user can decide to use the same system or implement a new one from scratch. The latter is

> outside the scope of this document and is explained int he dedicated [`Weight`]() conceptual

Reviewer 2: I think we should bring your explanation into this document, since it makes sense as a developer doc.

document.

### Using the Default Weight

> The default weight annotation is beyond _simple_. Substrate, by default, uses _fixed_ weights and

Reviewer 2: ```suggestion
Substrate, by default, uses _fixed_ weights and
```

Please avoid these kinds of sayings.

the enum representing them is as follows:

```rust
pub enum SimpleDispatchInfo {
    /// A normal dispatch with fixed weight.
    FixedNormal(Weight),
    /// A normal dispatch with the maximum weight.
    MaxNormal,
    /// A normal dispatch with no weight.
    FreeNormal,
    /// An operational dispatch with fixed weight.
    FixedOperational(Weight),
    /// An operational dispatch with the maximum weight.
    MaxOperational,
    /// An operational dispatch with no weight.
    FreeOperational,
}
```

This enum groups all dispatches into _normal_ and _operational_ (which makes the
implementation of `ClassifyDispatch` pretty trivial) and gives them a fixed Weight. Fixed in this
context means that the arguments of the dispatch do not play any role in the weight. Dispatches
classified as _operational_ are exempt from paying both the `base_fee`, and `length_fee`.

A simple example of using this enum in your runtime is:

> ```rust

Reviewer 2: I would prefer that we use subheadings and separate out these different kinds of fees.

Then it would show up in table of contents, and would allow direct linking.

Comments can become normal text at this point. I can make this change at a later point.

Reviewer: I don't get the suggestion: make the code a subheading, or `A simple example of using this enum in your runtime is:` a new one? 

yeah I don't mind you or anyone applying changes. I did my best to put the content right and if it needs reshaping then it is fine.

use sr_primitives::weights::{SimpleDispatchInfo};

decl_module! {
    // This means that this function has no weight. It will not contribute to block fullness at all,
    // and no weight-fee is applied.
    #[weight = SimpleDispatchInfo::FreeNormal]
    pub fn some_normal_function_light() { noop(); }

    // This function will always have a weight `10`.
    #[weight = SimpleDispatchInfo::FixedNormal(10)]
    pub fn some_normal_function_heavy() { some_computation(); }

    // This function will have a fixed weight but can consume the reserved operational portion as well.
    #[weight = SimpleDispatchInfo::FixedOperational(20)]
    pub fn mission_critical_function() { some_sudo_op(); }

    // This will automatically get `#[weight = SimpleDispatchInfo::default()]`
    pub fn something_else
}
```

> **Be careful!** The default implementation of `SimpleDispatchInfo` resolves to

Reviewer 2: ```suggestion
> **Note:** Be careful! The default implementation of `SimpleDispatchInfo` resolves to
```

---
title: Transaction Fees
---

When transactions are submitted to a blockchain, they are executed by the nodes
in the network. To be economically sustainable, nodes charge a fee to execute a
transaction. This fee must be covered by the submitter of the transaction. The
cost to execute transactions could vary over orders of magnitude, and thus
Substrate provides a flexible mechanism for characterizing the total, presumably
minimum, cost of a transaction in order to be included into a block.

> The fee system is heavily linked to the [weight system](). Make sure to read and

Reviewer 2: ```suggestion
The fee system is heavily linked to the [weight system](conceptual/runtime/weight.md). Make sure to read and
```

understand weights section before continuing this document.

## Transaction Fees

The fee to include a transaction consists of three parts:

* `base_fee` a fixed fee that is applied to every single transaction. See

>   [`TransactionBaseFee`](https://crates.parity.io/srml_transaction_payment/trait.Trait.html#associatedtype.TransactionBaseFee)

Reviewer 2: ```suggestion
  [`TransactionBaseFee`](/rustdocs/master/srml_transaction_payment/trait.Trait.html#associatedtype.TransactionBaseFee)
```

* `length_fee` a per-byte fee that is multiplied by the length, in bytes, of the
  encoded transaction. See

>   [`TransactionByteFee`](https://crates.parity.io/srml_transaction_payment/trait.Trait.html#associatedtype.TransactionByteFee)

Reviewer 2: ```suggestion
  [`TransactionByteFee`](/rustdocs/master/srml_transaction_payment/trait.Trait.html#associatedtype.TransactionByteFee)
```

associated type in `transaction-payment` is defined as a `Convert<Fixed64,
Fixed64>`, which should be read as: it receives the previous multiplier and
spits out the next one.

The default update function is inspired by the Polkadot network and implements a
targeted adjustment in which a target saturation level of block weight is
defined. If the previous block is more saturated, then the fees are slightly
increases. Similarly, if less transaction than the target are in the previous
block, fees are decreased by a small amount. More information about this can be
found in the [web3 research
page](https://research.web3.foundation/en/latest/polkadot/Token%20Economics/#relay-chain-transaction-fees).

## Substrate Weight System

> This section assumes that you have already read the `Weight` section

The entire substrate runtime module library is already annotated with a simple
and fixed weight system. A user can decide to use the same system or implement a
new one from scratch. The latter is outside the scope of this document and is
explained int he dedicated [`Weight`]() conceptual document.

> ### Using the Default Weight

Reviewer: yes I would leave this open and we can apply it at the end

> Transactors may also include optional tips in transactions to incentivize validators to include the transactions faster. Tips are a separate concept and are not covered here.

## Adjusting Multipliers
The `base_fee` and `byte_fee` are constants in the Transaction Fees module. The weight multiplier seems to adjust dynamically. @kainenigma can you explain that.

## Weight Basics
Transaction weights allow developers to express the cost of executing a transaction as a function of it's arguments. Some theoretical and practical advice for doing so is given below.

## Dispatch Parameters
Every transaction is classified as either Normal
FixedNormal
FreeNormal
MaxNormal

> Here we've presented the Normal variants. There is also `FixedOperational` and its relatives. Explain the difference and decide whether this article should be about dispatch info more generally.

## Examples
O(n) in an integer parameter
link to recipes

## Limitations
some operations get more expensive as storage grows and shrinks.
Is the weighting function open to change through governance short of a runtime upgrade?

## What is the default weight if left undefined?

> I think it's `FreeNormal`

Reviewer 2: From https://github.com/paritytech/substrate/blob/master/core/sr-primitives/src/weights.rs

```
impl Default for SimpleDispatchInfo {
	fn default() -> Self {
		// Default weight of all transactions.
		SimpleDispatchInfo::FixedNormal(10_000)
	}
}
```

submitter of the transaction. The cost to execute transactions could vary over orders of magnitude,
and thus Substrate provides a flexible mechanism for characterizing the total, presumably minimum,
cost of a transaction in order to be included into a block.

Furthermore, a number of resources in any chain can be limited by nature such as time, memory and
computation. A mechanism should exist to prevent individual resource-consuming components of the
chain, including dispatchables, to consume too much of any of the mentioned resources. This concept
is represented in substrate by weights.

## Weights Basics

The fee system is dependent on the weight system. Hence, we will first briefly describe the weights.

As mentioned, weights can technically refer to, or represent, numerous _limited_ resources. However,
at the time of this writing, substrate weights are simply a numeric value (todo: link). Each
dispatchable function is given a weight using the `#[weight = $x]` annotation, where `$x` is some
function capable of determining the weight of the dispatch. `$x` can access and examine the
arguments of the dispatch. Nonetheless, a weight calculation should always:

> - Be computable __ahead of dispatch__. A block producer, or a validator, should be able to examine

Author: So this is why it doesn't make sense to use storage values in the calculation.

Reviewer: Yes I have not made that more clear.
